File: sdk/src/main/java/org/coralibre/android/sdk/internal/database/persistent/entity/EntityDiagnosisKey.java
package org.coralibre.android.sdk.internal.database.persistent.entity;

import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;

import org.coralibre.android.sdk.internal.datatypes.DiagnosisKey;

import java.util.LinkedList;
import java.util.List;

import static androidx.room.ForeignKey.CASCADE;

@Entity(foreignKeys = @ForeignKey(
    entity = EntityToken.class,
    parentColumns = "tokenString",
    childColumns = "tokenString",
    onDelete = CASCADE
))
public class EntityDiagnosisKey {

    /**
     * This field is not used. It is only here to give the database a primary key.
     */
    @PrimaryKey(autoGenerate = true)
    public long dbPrimaryKey;

    public String tokenString;

    public byte[] keyData;
    public long intervalNumber;
    public int transmissionRiskLevel;


    public EntityDiagnosisKey() {}

    public EntityDiagnosisKey(String token, final DiagnosisKey diagnosisKey) {
        tokenString = token;
        keyData = diagnosisKey.getKeyData();
        intervalNumber = diagnosisKey.getInterval().get();
        transmissionRiskLevel = diagnosisKey.getTransmissionRiskLevel();
    }


    public static List<EntityDiagnosisKey> toEntityDiagnosisKeys(String token, List<DiagnosisKey> diagKeys) {
        LinkedList<EntityDiagnosisKey> result = new LinkedList<EntityDiagnosisKey>();
        for (DiagnosisKey dk : diagKeys) {
            result.add(new EntityDiagnosisKey(token, dk));
        }
        return result;
    }

}

File: sdk/src/main/java/org/coralibre/android/sdk/internal/database/persistent/entity/EntityDiagnosisKey.kt
package org.coralibre.android.sdk.internal.database.persistent.entity

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey
import org.coralibre.android.sdk.internal.datatypes.DiagnosisKey

/**
 * @param dbPrimaryKey This field is not used. It is only here to give the database a primary key.
 */
@Entity(
    foreignKeys = [
        ForeignKey(
            entity = EntityToken::class,
            parentColumns = ["tokenString"],
            childColumns = ["tokenString"],
            onDelete = ForeignKey.CASCADE
        )
    ]
)
data class EntityDiagnosisKey @JvmOverloads constructor(
    val tokenString: String,
    val keyData: ByteArray,
    val intervalNumber: Long,
    val transmissionRiskLevel: Int,
    @PrimaryKey(autoGenerate = true)
    val dbPrimaryKey: Long = 0,
) {
    constructor(token: String, diagnosisKey: DiagnosisKey) : this(
        tokenString = token,
        keyData = diagnosisKey.keyData,
        intervalNumber = diagnosisKey.interval.get(),
        transmissionRiskLevel = diagnosisKey.transmissionRiskLevel,
    )

    companion object {
        @JvmStatic
        fun toEntityDiagnosisKeys(
            token: String,
            diagKeys: List<DiagnosisKey>
        ): List<EntityDiagnosisKey> {
            return diagKeys.toEntityDiagnosisKeys(token)
        }
    }
}

fun List<DiagnosisKey>.toEntityDiagnosisKeys(token: String) = map { EntityDiagnosisKey(token, it) }
